src/sentry/monitors/system_incidents.py

```diff
@@ -166,6 +177,258 @@ def get_clock_tick_volume_metric(tick: datetime) -> float | None:
+    # anomalous state, mark this tick as anomalous.
+    if last_decision == Decision.NORMAL and metric_decisions[-1] == Decision.ABNORMAL:
+        return make_decision(Decision.ABNORMAL, AnomalyTransition.ABNORMALITY_STARTED)
+
+    # RECOVERED from abnormality
+    #
+    # If the previous result was anomalous check and if we have recovered and can
+    # backfill these decisions as normal
+    if last_decision == Decision.ABNORMAL and all(metric_matches(Decision.NORMAL)):
+        _backfill_decisions(past_ts, Decision.NORMAL, Decision.ABNORMAL)
+        return make_decision(Decision.NORMAL, AnomalyTransition.ABNORMALITY_RECOVERED)
+
+    # RECOVERING from an incident
+    #
+    # If we are actively in an incident and the most recent metric decision has
+    # recovered to normal we can de-escalate the incident to abnormal.
+    if last_decision == Decision.INCIDENT and metric_decisions[-1] == Decision.NORMAL:
+        return make_decision(Decision.RECOVERING, AnomalyTransition.INCIDENT_RECOVERING)
+
+    # RECOVERED from an incident
+    #
+    # If the previous result was anomalous check if we have recovered and can
+    # backfill these decisions as normal
+    if last_decision == Decision.RECOVERING and all(metric_matches(Decision.NORMAL)):
+        _backfill_decisions(past_ts, Decision.NORMAL, Decision.RECOVERING)
+        return make_decision(Decision.NORMAL, AnomalyTransition.INCIDENT_RECOVERED)
+
+    # FAILED RECOVERY from an incident
+    #
+    # If an incident had begun recovering but we've detected an abnormality
+    # again backfill all recovery decisions to an incident decision.
+    if last_decision == Decision.RECOVERING and metric_decisions[-1] == Decision.ABNORMAL:
+        _backfill_decisions(past_ts, Decision.INCIDENT, Decision.RECOVERING)
+        return make_decision(Decision.INCIDENT, AnomalyTransition.INCIDENT_STARTED)
+
+    # ENTERING an incident
+    #
+    # If we're not already in an incident and the most recent metric decision
+    # is an incident, mark this tick as an incident and backfill all abnormal
+    # decisions to an incident decision.
+    if last_decision != Decision.INCIDENT and metric_decisions[-1] == Decision.INCIDENT:
+        _backfill_decisions(past_ts, Decision.INCIDENT, Decision.ABNORMAL)
+        return make_decision(Decision.INCIDENT, AnomalyTransition.INCIDENT_STARTED)
+
+    # No decision transition. Use the previous decision
+    return make_decision(last_decision)
+
+
+def get_clock_tick_decision(tick: datetime) -> TickAnomalyDecision | None:
+    """
+    Retrieve the TickAnomalyDecision for a specific clock tick.
+    """
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    if value := redis_client.get(MONITOR_TICK_DECISION.format(ts=_make_reference_ts(tick))):
+        return TickAnomalyDecision.from_str(value)
+    else:
+        return None
+
+
+def _backfill_decisions(
```

Comment:
It might be worth fetching the keys in batches just to cut down a bit on IO. 

The infinite loop is a concern... maybe we should set some really high upper bound?

1440 minutes in a day, so if we set this to 10k then we'd likely never hit it unless we had an actual bug. If we're down for over a day then we have bigger problems

Reply:
Yeah. I like that idea

Reply:
Let me deal with this in a follow up PR actually if that's OK?